includes/class-wc-stripe-intent-controller.php

```diff
@@ -604,6 +606,62 @@ public function update_order_status_ajax() {
 	}
 
 	/**
+	 * Handle AJAX request if error occurs while confirming intent.
+	 * We will log the error and update the order.
+	 *
+	 * @throws WC_Stripe_Exception
+	 */
+	public function update_failed_order_ajax() {
+		try {
+			$is_nonce_valid = check_ajax_referer( 'wc_stripe_update_failed_order_nonce', false, false );
+			if ( ! $is_nonce_valid ) {
+				throw new WC_Stripe_Exception( 'missing-nonce', __( 'CSRF verification failed.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$order_id  = isset( $_POST['order_id'] ) ? absint( $_POST['order_id'] ) : null;
+			$intent_id = isset( $_POST['intent_id'] ) ? wc_clean( wp_unslash( $_POST['intent_id'] ) ) : '';
+			$order     = wc_get_order( $order_id );
+
+			if ( ! empty( $order_id ) && ! empty( $intent_id ) && is_object( $order ) ) {
+				$payment_needed = 0 < $order->get_total();
+				if ( $payment_needed ) {
+					$intent = WC_Stripe_API::retrieve( "payment_intents/$intent_id" );
+				} else {
+					$intent = WC_Stripe_API::retrieve( "setup_intents/$intent_id" );
+				}
+				$error = $intent->last_payment_error;
+
+				if ( ! empty( $error ) ) {
+					WC_Stripe_Logger::log( 'Error when processing payment: ' . $error->message );
+					throw new WC_Stripe_Exception( __( "We're not able to process this payment. Please try again later.", 'woocommerce-gateway-stripe' ) );
+				}
+
+				// Use the last charge within the intent to proceed.
+				$gateway = $this->get_gateway();
+				if ( isset( $intent->charges ) ) {
```

Comment:
This fallback area is a little awkward just because I'm not even sure how it would be reached in a real world scenario, but I like that it is at least there! To test this area, I just made the failed order call before even attempting to confirm the intent. This led to the creation of an order on-hold, but it didn't have any charge or payment intent IDs attached to allow the merchant to manually process the charge if desired. What I found was that we should also be checking that `$intent->charges->data` is not empty before using the charge in the `process_response` call instead of the intent.

Long story short, I think the conditional should have: 
```suggestion
				if ( isset( $intent->charges ) && ! empty( $intent->charges->data ) ) {
```

Reply:
Cool, suggestion seems sound: committed. As stated, I'm also not sure how to get to that particular section of code without any wizardry, but you know what they say about death, taxes, and Stripe doing strange things unexpectedly...